File: frontend/src/types/portfolio.ts
// src/types/portfolio.ts

/**
 * Portfolio interface representing a user's investment portfolio
 */
export interface Portfolio {
    id: number;
    name: string;
    description?: string;
    portfolio_type: PortfolioType;
    initial_capital: number;
    current_value: number;
    cash_balance: number;
    is_active: boolean;
    broker?: string;
    broker_account_id?: string;
    created_at: string; // ISO date string
    updated_at?: string; // ISO date string
    user_id: number;
    positions?: Position[];
  }
  
  /**
   * Possible portfolio types
   */
  export type PortfolioType = 'paper' | 'live';
  
  /**
   * Position interface representing a holding in a portfolio
   */
  export interface Position {
    id: number;
    symbol: string;
    quantity: number;
    average_entry_price: number;
    current_price?: number;
    current_value?: number;
    unrealized_pnl?: number;
    unrealized_pnl_pct?: number;
    is_open: boolean;
    opened_at: string; // ISO date string
    closed_at?: string; // ISO date string
    strategy_id?: number;
    portfolio_id: number;
    created_at: string; // ISO date string
    updated_at?: string; // ISO date string
  }
  
  /**
   * Trade interface representing an executed trade in a portfolio
   */
  export interface Trade {
    id: number;
    symbol: string;
    trade_type: TradeType;
    quantity: number;
    price: number;
    total_amount: number;
    commission: number;
    status: TradeStatus;
    executed_at?: string; // ISO date string
    realized_pnl?: number;
    realized_pnl_pct?: number;
    notes?: string;
    order_id?: string;
    user_id: number;
    portfolio_id: number;
    strategy_id?: number;
    created_at: string; // ISO date string
    updated_at?: string; // ISO date string
  }
  
  /**
   * Trade types
   */
  export type TradeType = 'buy' | 'sell' | 'short' | 'cover';
  
  /**
   * Trade status types
   */
  export type TradeStatus = 'pending' | 'filled' | 'canceled' | 'rejected' | 'partial';
  
  /**
   * Interface for portfolio creation request
   */
  export interface PortfolioCreateRequest {
    name: string;
    description?: string;
    portfolio_type: PortfolioType;
    initial_capital: number;
    broker?: string;
    broker_account_id?: string;
  }
  
  /**
   * Interface for portfolio update request
   */
  export interface PortfolioUpdateRequest {
    name?: string;
    description?: string;
    is_active?: boolean;
    broker?: string;
    broker_account_id?: string;
  }
  
  /**
   * Interface for position creation request
   */
  export interface PositionCreateRequest {
    symbol: string;
    quantity: number;
    average_entry_price: number;
    strategy_id?: number;
  }
  
  /**
   * Interface for trade creation request
   */
  export interface TradeCreateRequest {
    symbol: string;
    trade_type: TradeType;
    quantity: number;
    price: number;
    commission?: number;
    notes?: string;
    strategy_id?: number;
  }
  
  /**
   * Interface for portfolio performance data
   */
  export interface PortfolioPerformance {
    portfolio_id: number;
    initial_value: number;
    current_value: number;
    total_return: number;
    total_return_pct: number;
    performance_data: PerformancePoint[];
  }
  
  /**
   * Interface for a point in the performance data
   */
  export interface PerformancePoint {
    date: string; // ISO date string
    value: number;
  }

File: backend/models/portfolio.py
# backend/models/portfolio.py
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, ForeignKey, JSON, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum

from backend.models import Base

class PortfolioType(str, enum.Enum):
    PAPER = "paper"
    LIVE = "live"

class Portfolio(Base):
    __tablename__ = "portfolios"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    description = Column(String)
    
    # Portfolio type (paper or live trading)
    portfolio_type = Column(Enum(PortfolioType), default=PortfolioType.PAPER, nullable=False)
    
    # Portfolio metadata
    initial_capital = Column(Float, nullable=False)
    current_value = Column(Float)
    cash_balance = Column(Float)
    is_active = Column(Boolean, default=True)
    
    # Connected broker (if any)
    broker = Column(String)
    broker_account_id = Column(String)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    user_id = Column(Integer, ForeignKey("users.id"))
    user = relationship("User", back_populates="portfolios")
    positions = relationship("Position", back_populates="portfolio", cascade="all, delete-orphan")
    trades = relationship("Trade", back_populates="portfolio")
    
    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "portfolio_type": self.portfolio_type,
            "initial_capital": self.initial_capital,
            "current_value": self.current_value,
            "cash_balance": self.cash_balance,
            "is_active": self.is_active,
            "broker": self.broker,
            "broker_account_id": self.broker_account_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "user_id": self.user_id,
            "positions": [position.to_dict() for position in self.positions] if self.positions else []
        }

class Position(Base):
    __tablename__ = "positions"

    id = Column(Integer, primary_key=True, index=True)
    symbol = Column(String, nullable=False)
    quantity = Column(Float, nullable=False)
    average_entry_price = Column(Float, nullable=False)
    
    # Current position data
    current_price = Column(Float)
    current_value = Column(Float)
    unrealized_pnl = Column(Float)
    unrealized_pnl_pct = Column(Float)
    
    # Position metadata
    is_open = Column(Boolean, default=True)
    opened_at = Column(DateTime(timezone=True), server_default=func.now())
    closed_at = Column(DateTime(timezone=True))
    
    # Associated strategy (optional)
    strategy_id = Column(Integer, ForeignKey("strategies.id"), nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    portfolio_id = Column(Integer, ForeignKey("portfolios.id"))
    portfolio = relationship("Portfolio", back_populates="positions")
    strategy = relationship("Strategy")
    
    def to_dict(self):
        return {
            "id": self.id,
            "symbol": self.symbol,
            "quantity": self.quantity,
            "average_entry_price": self.average_entry_price,
            "current_price": self.current_price,
            "current_value": self.current_value,
            "unrealized_pnl": self.unrealized_pnl,
            "unrealized_pnl_pct": self.unrealized_pnl_pct,
            "is_open": self.is_open,
            "opened_at": self.opened_at.isoformat() if self.opened_at else None,
            "closed_at": self.closed_at.isoformat() if self.closed_at else None,
            "strategy_id": self.strategy_id,
            "portfolio_id": self.portfolio_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None
        }